### DISCUSSION TITLE
rbenv install does not work on mac M1

### DISCUSSION THREAD

**Discussion by OrkunTokdemir** [rbenv/ruby-build#1922] [Build failures]
I installed `rbenv` today but it returns a couple of errors when I tried to run `rbenv install 3.1.0`

My Setup 
![image](https://user-images.githubusercontent.com/21346859/151032727-39a34da6-46c6-46f1-a312-5088aae80a00.png)

```
Darwin orkuns-MacBook-Pro.local 21.2.0 Darwin Kernel Version 21.2.0: Sun Nov 28 20:29:10 PST 2021; root:xnu-8019.61.5~1/RELEASE_ARM64_T8101 arm64
```

### iterm output 
```
$ rbenv install 3.0.3
Downloading ruby-3.0.3.tar.gz...
-> https://cache.ruby-lang.org/pub/ruby/3.0/ruby-3.0.3.tar.gz
Installing ruby-3.0.3...
ruby-build: using readline from homebrew

BUILD FAILED (macOS 12.1 using ruby-build 20211227)

Inspect or clean up the working tree at /var/folders/9v/pqhg0pdx5n10b1xh39g989kc0000gp/T/ruby-build.20220125185613.60847.G3zm2f
Results logged to /var/folders/9v/pqhg0pdx5n10b1xh39g989kc0000gp/T/ruby-build.20220125185613.60847.log

Last 10 log lines:
compiling nofree.c
compiling normalize.c
compiling qsort.c
compiling rb_interned_str.c
compiling rb_str_dup.c
compiling set_len.c
linking shared-object cgi/escape.bundle
linking shared-object coverage.bundle


linking shared-object -test-/string.bundle
make: *** [build-ext] Error 2
```
### Error details 
```
In file included from ./bigdecimal.h:14:
./missing.h:127:1: error: static declaration of 'rb_rational_num' follows non-static declaration
rb_rational_num(VALUE rat)
^
../.././include/ruby/internal/intern/rational.h:128:7: note: previous declaration is here
VALUE rb_rational_num(VALUE rat);
      ^
In file included from bigdecimal.c:13:
In file included from ./bigdecimal.h:14:
./missing.h:139:1: error: static declaration of 'rb_rational_den' follows non-static declaration
rb_rational_den(VALUE rat)
^
../.././include/ruby/internal/intern/rational.h:138:7: note: previous declaration is here
VALUE rb_rational_den(VALUE rat);
      ^
In file included from bigdecimal.c:13:
In file included from ./bigdecimal.h:14:
./missing.h:153:1: error: static declaration of 'rb_complex_real' follows non-static declaration
rb_complex_real(VALUE cmp)
^
../.././include/ruby/internal/intern/complex.h:101:7: note: previous declaration is here
VALUE rb_complex_real(VALUE z);
      ^
In file included from bigdecimal.c:13:
In file included from ./bigdecimal.h:14:
./missing.h:165:1: error: static declaration of 'rb_complex_imag' follows non-static declaration
rb_complex_imag(VALUE cmp)
^
../.././include/ruby/internal/intern/complex.h:110:7: note: previous declaration is here
VALUE rb_complex_imag(VALUE z);
      ^
In file included from bigdecimal.c:13:
In file included from ./bigdecimal.h:14:
./missing.h:189:1: error: redefinition of 'rb_array_const_ptr'
rb_array_const_ptr(VALUE a)
^
../.././include/ruby/internal/core/rarray.h:393:1: note: previous definition is here
rb_array_const_ptr(VALUE a)
^
In file included from bigdecimal.c:13:
In file included from ./bigdecimal.h:14:
./missing.h:208:1: error: static declaration of 'rb_sym2str' follows non-static declaration
rb_sym2str(VALUE sym)
^
../.././include/ruby/internal/symbol.h:211:7: note: previous declaration is here
VALUE rb_sym2str(VALUE id);
```
I also added the log file 
[ruby-build.20220125201815.95010.log](https://github.com/rbenv/rbenv/files/7936565/ruby-build.20220125201815.95010.log)

**Comment by notjames** [comment]
Happening to me, too, except I'm running a MBP 2019 with Big Sur attempting to install Ruby 3.1.0 with `rbenv`. It's not clear to me which dev-lib I'm missing to fix this issue.

**Reply by notjames** [reply]
TL;DR 

I had to give up on using `rbenv` so I `brew uninstall`ed everything related to ruby (except `ruby-build`). Then I ran `brew install ruby@3.0` (after doing a `brew upgrade`, which at least got me a ruby version > 2.x.

**Reply by mislav** [reply]
@notjames Sorry for the trouble! If you ever want to give ruby-build another try, please see if https://github.com/rbenv/ruby-build/issues/1505#issuecomment-752274233 helps in your case.

**Reply by mislav** [reply]
Also, please check that you do not have `binutils` installed through macports https://trac.macports.org/ticket/62624#comment:6

**Reply by notjames** [reply]
I actually did try the steps in #1505 but to no avail. I haven't tried the second point yet, though. I think I do in fact have `binutils` installed through `brew`. Does that matter?

**Reply by mislav** [reply]
@notjames No, having them installed via Homebrew should be fine, since Homebrew will never link them to `/usr/local` by default.
```
==> Caveats
binutils is keg-only, which means it was not symlinked into /usr/local,
because Apple's CLT provides the same tools.
```

**Reply by notjames** [reply]
ok, I was able to get back to this today and decided to poke around again. Turns out that this is not an `rbenv` issue. It's a `ruby-build` by way of `ruby` issue gleaned from a discussion thread [here](https://github.com/rbenv/ruby-build/issues/1505#issuecomment-751834144) which lead to the fix [here](https://bugs.ruby-lang.org/issues/17486) by applying a patch derived from [here](https://github.com/ruby/ruby/commit/ed01cc8fdc845b1db258eb3e35b3ba834463672e)

To create the patch I had to run a few commands first and massage some things to make this work with `rbenv` so I'll provide my patch here and then sum up the best I can...

the patch:
```
--- /var/tmp/configure.ac.diff  2022-06-25 13:10:43.641427516 -0600
+++ configure.ac        2022-04-12 05:11:15.000000000 -0600
@@ -207,6 +207,13 @@
 ],
 [aix*], [
     AC_PATH_TOOL([NM], [nm], [/usr/ccs/bin/nm], [/usr/ccs/bin:$PATH])
+],
+[darwin*], [
+    AS_IF([libtool 2>&1 | grep no_warning_for_no_symbols > /dev/null], [
+        ac_cv_prog_ac_ct_RANLIB=:
+        ac_cv_prog_ac_ct_AR='libtool -static'
+        rb_cv_arflags='-no_warning_for_no_symbols -o'
+    ])
 ])
 AS_CASE(["${target_os}"],
 [cygwin*|msys*|mingw*], [
```

copy and paste that into `/var/tmp/configure.ac.patch`
then do:

```
$ brew upgrade
$ brew install rbenv ruby-build libffi  < and anything else you need here >
$ < /var/tmp/configure.ac.patch rbenv install -p 3.1.2
```

This finally worked for me:
```
at 15:01:37 ❯ < /var/tmp/configure.ac.patch rbenv install -p 3.1.2
Downloading ruby-3.1.2.tar.gz...
-> https://cache.ruby-lang.org/pub/ruby/3.1/ruby-3.1.2.tar.gz
Installing ruby-3.1.2...
patching file configure.ac
ruby-build: using readline from homebrew
Installed ruby-3.1.2 to /Users/jimconner/.rbenv/versions/3.1.2
```

**Reply by matheuuss** [reply]
Thank you @notjames !! The above solution worked for me, only that the configure.ac file was in src/ruby-3.3.1.
